### PR TITLE
Move Octopus.Shared.Contracts to Octopus.Tentacle.Contracts

### PR DESCRIPTION
**Pull request by tothegills**
We would like to publish the Tentacle script contracts from the Tentacle solution (rather than having copies of them everywhere). In order to maintain backwards compatibility, the contracts either needs to stay in the `Octopus.Shared` assembly or pretend they are in the `Octopus.Shared` assembly. If we keep them in the `Octopus.Shared` assembly, that assembly will clash with the old `Octopus.Shared` when imported in projects. So, we will pretend.

This PR implements custom JSON binders to map to and from the new assembly.
Relates to https://github.com/OctopusDeploy/ResearchAndDevelopment/issues/339
Server build: https://github.com/OctopusDeploy/OctopusDeploy/pull/13878

### CODE REVIEW

File: source/Octopus.Tentacle/Communications/AssemblyMappingSerializationBinderDecorator.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using Newtonsoft.Json.Serialization;
+
+namespace Octopus.Tentacle.Communications
+{
+    public class AssemblyMappingSerializationBinderDecorator : ISerializationBinder
```
Comment (tothegills) [inline source/Octopus.Tentacle/Communications/AssemblyMappingSerializationBinderDecorator.cs:6]: This is important - maps from the old assembly name (Octopus.Shared) to the new assembly name (Octopus.Tentacle.Contracts).

File: source/Octopus.Tentacle/Communications/NamespaceMappingSerializationBinderDecorator.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using Newtonsoft.Json.Serialization;
+
+namespace Octopus.Tentacle.Communications
+{
+    public class NamespaceMappingSerializationBinderDecorator : ISerializationBinder
```
Comment (tothegills) [inline source/Octopus.Tentacle/Communications/NamespaceMappingSerializationBinderDecorator.cs:6]: This is important - maps from the old namespace (Octopus.Shared.Contracts) to the new namespace (Octopus.Tentacle.Contracts).

File: source/Octopus.Tentacle/Communications/TentacleCommunicationsModule.cs [hunk 3]
```diff
@@ -20,7 +20,16 @@ protected override void Load(ContainerBuilder builder)
             {
                 var configuration = c.Resolve<ITentacleConfiguration>();
                 var services = c.Resolve<IServiceFactory>();
-                var halibutRuntime = new HalibutRuntimeBuilder().WithServiceFactory(services).WithServerCertificate(configuration.TentacleCertificate).Build();
+                var halibutRuntime = new HalibutRuntimeBuilder()
+                    .WithServiceFactory(services)
+                    .WithServerCertificate(configuration.TentacleCertificate)
+                    .WithMessageSerializer(serializerBuilder => serializerBuilder.WithSerializerSettings(settings =>
```
Comment (tothegills) [inline source/Octopus.Tentacle/Communications/TentacleCommunicationsModule.cs:26]: Register the new binders to performing the mapping to/from assembly and namespace.

File: source/Octopus.Tentacle/Time/TimeModule.cs [hunk 4]
```diff
@@ -2,7 +2,7 @@
 using Autofac;
 using Octopus.Time;
 
-namespace Octopus.Shared.Time
+namespace Octopus.Tentacle.Time
```
Comment (tothegills) [inline source/Octopus.Tentacle/Time/TimeModule.cs:5]: Picked up some extra namespace corrections.

File: source/Octopus.Tentacle.Contracts/Octopus.Tentacle.Contracts.csproj [hunk 5]
```diff
@@ -0,0 +1,30 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <RootNamespace>Octopus.Tentacle.Contracts</RootNamespace>
+    <AssemblyName>Octopus.Tentacle.Contracts</AssemblyName>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+    <OutputPath>bin\</OutputPath>
+    <TargetFrameworks>net452;netcoreapp3.1</TargetFrameworks>
```
Comment (ankithkonda) [inline source/Octopus.Tentacle.Contracts/Octopus.Tentacle.Contracts.csproj:8]: Should these be updated as they're EOL?

File: source/Octopus.Tentacle.Contracts/Legacy/AssemblyMappingSerializationBinderDecorator.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using Newtonsoft.Json.Serialization;
+
+namespace Octopus.Tentacle.Contracts.Legacy
+{
+    public class AssemblyMappingSerializationBinderDecorator : ISerializationBinder
```
Comment (ankithkonda) [inline source/Octopus.Tentacle.Contracts/Legacy/AssemblyMappingSerializationBinderDecorator.cs:6]: TIL ISerializationBinder, nice 👍

**Review by ankithkonda — APPROVED** [review]
All makes sense to me 👍

File: source/Octopus.Tentacle.Contracts/Octopus.Tentacle.Contracts.csproj (hunk 5 above)
Comment (tothegills) [reply source/Octopus.Tentacle.Contracts/Octopus.Tentacle.Contracts.csproj:8]: Probably yes, but that is a separate discussion about what we are willing to support.

**Pull request merged**